src/jshint.js

```diff
@@ -4909,14 +4704,14 @@ var JSHINT = (function() {
             return true;
           // When we are in the "generate" state of the list comp,
           } else if (_current && _current.mode === "generate") {
-            isundef(funct, "W117", state.tokens.curr, v);
+            funct["(scope)"].current.use(v, state.tokens.curr);
             return true;
           // When we are in "filter" state,
           } else if (_current && _current.mode === "filter") {
             // we check whether current variable has been declared
             if (use(v)) {
               // if not we warn about it
-              isundef(funct, "W117", state.tokens.curr, v);
+              funct["(scope)"].current.use(v, state.tokens.curr);
```

Comment:
Added tests and discovered 2 more existing bugs (1 fixed, one not) - #2439 #2440